### PR TITLE
Refactors Linux agent install to handle existing installation

### PR DESCRIPTION
**Pull request by eemperor**

### CODE REVIEW

**Comment by lorengordon** [comment]
@eemperor Looks like this project needs some updates to the CI configs...

**Comment by eemperor** [comment]
Yeah.  It's been a long time since this project has been updated.

File: mcafee-agent/elx/init.sls [hunk 1]
```diff
@@ -67,6 +67,8 @@ Install McAfee Agent:
     - cwd: '/root'
     - require:
       - file: Stage McAfee Install Archive
+    - success_retcodes:
+      - 1
```
Comment (lorengordon) [inline mcafee-agent/elx/init.sls]: Isn't this supposed to be a `2`?
```suggestion
      - 2
```

Comment (eemperor) [reply mcafee-agent/elx/init.sls]: It's `1` in this case.

`[ERROR   ][2625] {'pid': 2750, 'retcode': 1, 'stdout': 'An higher or same version of the agent is already installed. Installation cannot continue', 'stderr': ''}`

Comment (lorengordon) [reply mcafee-agent/elx/init.sls]: Oh... that's more of a stateful failure. Not good. If the installer has no mechanism for testing that the package is already installed and exiting without error, we need to handle that case explicitly and avoid calling the installer.

Comment (lorengordon) [reply mcafee-agent/elx/init.sls]: I think we need this feature, which hasn't been released yet. :( Looking like it'll be included in salt 3004...

https://github.com/saltstack/salt/pull/59841

Comment (lorengordon) [reply mcafee-agent/elx/init.sls]: Actually, I think we're trying to handle that scenario in a limited way here, skipping the install if already installed:

https://github.com/plus3it/mcafee-agent-formula/blob/e5ee0801f300a8aa2da645a2e5baf09f78458033/mcafee-agent/elx/init.sls#L70-L73

So it shouldn't be running the installer at all, I think?

Comment (eemperor) [reply mcafee-agent/elx/init.sls]: Okay.  Yeah, I don't see an option to check install status. I tried `-u` and it gives the same result.

```
]# sh ./install.sh -h
./install.sh: option requires an argument -- h
Usage: ./install.sh [-i|-u|-b|-p|-r|-R|-h  ]
-i : fresh install
-u : upgrade install
-b :Upgrade Agent only , no server info will be updated
-p : generate provisioning rpm/deb
-r : Use this option while installing/upgrading, to run macmnsvc as root user
-R : Set Relay Server ('IP1:Port1[;IP2:Port2;...]')
-h : show this help
```

Until your patch is incorporated, maybe I'll try to see if I can add a state to check for an existing install.

Comment (eemperor) [reply mcafee-agent/elx/init.sls]: > Actually, I think we're trying to handle that scenario in a limited way here, skipping the install if already installed:
> 
> https://github.com/plus3it/mcafee-agent-formula/blob/e5ee0801f300a8aa2da645a2e5baf09f78458033/mcafee-agent/elx/init.sls#L70-L73
> 
> So it shouldn't be running the installer at all, I think?

Right.  We're not even providing a rpm.

Comment (lorengordon) [reply mcafee-agent/elx/init.sls]: >Right. We're not even providing a rpm.

The installer.sh script installs the rpms. That `unless` block is supposed to check whether the resulting rpms are already installed. Perhaps the rpm names have changed?

Comment (eemperor) [reply mcafee-agent/elx/init.sls]: Oh.  I missed the default values in `map.jinja'.  Let me check what's on the system I'm using for testing.

Comment (eemperor) [reply mcafee-agent/elx/init.sls]: Yeah.  Somethings odd is going on with that `unless`...
```
# rpm -q MFEcma
MFEcma-5.7.1-116.x86_64
# rpm -q MFErt
MFErt-2.0-131.i686
```
The rpms are there.

Comment (lorengordon) [reply mcafee-agent/elx/init.sls]: Ok, yeah, then let's try to "fix" this using `unless` or `onlyif`. Their logic is always a little weird, so doesn't surprise me that we have a bug there.

https://docs.saltproject.io/en/latest/ref/states/requisites.html#altering-states

Comment (eemperor) [reply mcafee-agent/elx/init.sls]: It seems something is not right with the handling of `unless`.  I'm hard coding the command and it's still not working.  The `install.sh` is still being executed.
```
Install McAfee Agent:
  cmd.run:
    - name: 'sh /root/install.sh {{ mcafee.installer_opts }}'
    - cwd: '/root'
    - require:
      - file: Stage McAfee Install Archive
    - unless:
      - 'rpm --quiet -q MFEcma'
{%- for key_file in mcafee.key_files %}
      - 'test -s {{ mcafee.keystore_directory }}/{{ key_file }}'
{%- endfor %}
```

**Review by lorengordon — APPROVED** [review]
:shipit:

**Pull request merged**